kong/clustering/control_plane.lua

```diff
@@ -105,27 +108,179 @@ function _M.new(parent)
     clients = setmetatable({}, { __mode = "k", })
   }
 
-  return setmetatable(self, {
+  setmetatable(self, {
     __index = function(tab, key)
       return _M[key] or parent[key]
     end,
   })
+
+  self.plugins_map = {}
+
+  return self
 end
 
 
+local function invalidate_keys_from_config(config_plugins, keys)
+  if not config_plugins then
+    return
+  end
+
+  for _, t in ipairs(config_plugins) do
+    if t and t["config"] then
+      local config = t["config"]
+      local name = gsub(t["name"], "-", "_")
+
+      -- Handle Redis configurations (regardless of plugin)
+      if config.redis then
+        local config_plugin_redis = t["config"].redis
+        for _, key in ipairs(keys["redis"]) do
+          if config_plugin_redis[key] ~= nil then
+            config_plugin_redis[key] = nil
+          end
+        end
+      end
+
+      -- Handle fields in specific plugins
+      if keys[name] ~= nil then
+        for _, key in ipairs(keys[name]) do
+          if config[key] ~= nil then
+            config[key] = nil
+          end
+        end
+      end
+    end
+  end
+end
+
+local function dp_version_num(dp_version)
+  local base = 1000000000
+  local version_num = 0
+  for _, v in ipairs(utils.split(dp_version, ".", 4)) do
+    v = v:match("^(%d+)")
+    version_num = version_num + base * tonumber(v, 10) or 0
+    base = base / 1000
+  end
+
+  return version_num
+end
+-- for test
+_M._dp_version_num = dp_version_num
+
+local removal_fields = {
```

Comment:
I wonder if those should live inside a separate Lua file/module. Having it inside `control_plane.lua` seems to be really inconvenient for updating/patching.

Reply:
that was the goal yes, ideally we will want that seperate file to be auto generated but
i would like to make that in a seperate PR. and this one just is just to take what's already
there in EE.